### PR TITLE
chore: new apisix upstream controller

### PR DESCRIPTION
**Pull request by tokers**
* `pkg/crd` contains translators to translate resources from declarative configurations to the APISIX resources.
* refactor the data structure of ApisixUpstream and Upstream.
* refactor the ApisixUpstream controller.
* endpoints controller uses the correct service port rather than the targetPort to compose the APISIX Upstream name.

### CODE REVIEW

**Comment by membphis** [comment]
ci failed  :(

**Comment by huahu0213** [comment]
> ci failed :(

So sorry ... I will figure it out tomorrow

**Comment by tokers** [comment]
> ci failed :(

Fixed.

**Comment by tokers** [comment]
@gxthrj @membphis Please take a look when you have time :)

**Comment by gxthrj** [comment]
CI failed

**Comment by membphis** [comment]
I think we need more test cases

image

**Comment by tokers** [comment]
@membphis It's somewhat difficult to test the controller loop in unit test cases, we test it already in the e2e test suites.

Also, the reason that the coverage ratio reduces is we deleted some files and its test cases.

**Comment by membphis** [comment]
> Also, the reason that the coverage ratio reduces is we deleted some files and its test cases.

nice

**Comment by tokers** [comment]
> > Also, the reason that the coverage ratio reduces is we deleted some files and its test cases.
> 
> nice

And I created an issue #245 , will add more cases in the future.

File: pkg/ingress/controller/apisix_upstream.go [hunk 1]
```diff
@@ -15,202 +15,244 @@
...
-		apisixUpstreamList:   apisixUpstreamInformer.Lister(),
-		apisixUpstreamSynced: apisixUpstreamInformer.Informer().HasSynced,
-		workqueue:            workqueue.NewNamedRateLimitingQueue(workqueue.NewItemFastSlowRateLimiter(1*time.Second, 60*time.Second, 5), "ApisixUpstreams"),
-	}
-	apisixUpstreamInformer.Informer().AddEventHandler(
+func (c *Controller) newApisixUpstreamController() *apisixUpstreamController {
+	ctl := &apisixUpstreamController{
+		controller: c,
+		workqueue:  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ApisixUpstream"),
+		workers:    1,
```
[the remark this answers is not in the document]
Comment (tokers) [reply pkg/ingress/controller/apisix_upstream.go:42]: It's an extention, we can add good first issue, let others do it :)

[the remark this answers is not in the document]
Comment (gxthrj) [reply pkg/ingress/controller/apisix_upstream.go:42]: OK, please add a good first issue~

**Review by gxthrj — APPROVED** [review]

File: pkg/ingress/controller/apisix_upstream.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tokers) [reply pkg/ingress/controller/apisix_upstream.go:42]: OK, I'll.

**Pull request merged**